### PR TITLE
Don't use sleeps for synchronization

### PR DESCRIPTION
**Pull request by dryajov**
Previously, mplex relied on `sleepAsync` in order to synchronize reading from the remote connection into the bufferstream object, this was hiding a bug in mplex header decoding that would attempt to read a 0 length data body from the connection. 

This bug would cause intermittent hangings in various tests, but the biggest issues was that it would not await for data in the mplex handle loop, but instead go into a thigh loop, which would require inserting sleeps to prevent it from locking up, the sleeps generate a timer object in chronos which would ultimately bloat chronos internal timers heap that would lead to random timeouts and such.

### CODE REVIEW

File: libp2p/peerinfo.nim [hunk 1]
```diff
@@ -105,9 +105,13 @@ proc publicKey*(p: PeerInfo): Option[PublicKey] {.inline.} =
     result = some(p.privateKey.getKey())
 
 proc id*(p: PeerInfo): string {.inline.} =
-  result = p.peerId.pretty()
+  if not isNil(p):
```
Comment (dryajov) [inline libp2p/peerinfo.nim]: Not sure why p would be `Nil`, @cheatfate is this related to your changes by any chance?

Comment (dryajov) [reply libp2p/peerinfo.nim]: seems like this was related to an earlier change that has since been removed.

**Comment by dryajov** [comment]
@stefantalpalaru I'm getting weird but pretty consistent failures in CI but I can't reproduce them locally neither on mac nor linux. Are we running into os (fs, etc..) limits or is something else going on, any hints at where to look?

**Comment by stefantalpalaru** [comment]
I can reproduce it on Linux. Output from a binary compiled with `../../env.sh nim c -d:chronicles_log_level=TRACE tests/testnative.nim`: [log.txt.gz](https://github.com/status-im/nim-libp2p/files/4196036/log.txt.gz)

Adding `-d:release` makes the failure go away, but that's obviously not a fix.

File: libp2p/muxers/muxer.nim [hunk 2]
```diff
@@ -46,11 +46,14 @@ proc newMuxerProvider*(creator: MuxerConstructor, codec: string): MuxerProvider
 method init(c: MuxerProvider) =
   proc handler(conn: Connection, proto: string) {.async, gcsafe, closure.} =
     let muxer = c.newMuxer(conn)
+    var handlerFut: Future[void] = newFuture[void]()
```
Comment (arnetheduck) [inline libp2p/muxers/muxer.nim]:
```
let handlerFut =
  if ..: c.muxerHandler(..)
  else: dummy future
```

`let` is a safer construct (even like that, the code is wasteful when always calling `allFutures` - it also smells with `streamHandler` being set midway after `muxerHandler` is caled and `isNil` checked twice etc, classic hotspot for bugs)

Comment (dryajov) [reply libp2p/muxers/muxer.nim]: Thanks! Yes, this is just me moving things around trying to figure out where the raise is, this wasn't part of the original PR.

**Comment by dryajov** [comment]
Still running into what appear to be raises after removing the sleeps, this is still a WIP.

**Comment by sinkingsugar** [comment]
> Still running into what appear to be raises after removing the sleeps, this is still a WIP.

Is there anything I can do to help somehow?
floodsub is the failing one?

**Comment by dryajov** [comment]
> > Still running into what appear to be raises after removing the sleeps, this is still a WIP.
> 
> Is there anything I can do to help somehow?
> floodsub is the failing one?

It's a bit random, but right now it fails in floodsub, but removing floodsub tests doesn't actually make them pass, it just breaks in a different place.

**Comment by dryajov** [comment]
@sinkingsugar I was able to fix all the races and tests are passing now.
